EVE5: Theme Settings

alchucam opened the pull request:
#### What does this PR do?
- Theme Settings
User can go to the edit mode by clicking the setting icon and modify contents of various components. Currently supports edition of Banner color, BuildStatus JenkinsURL, BuildAF JenkinsURL, SlackChannel, Github Repopath.

- Minor Addition
Utilized Jest for simple unit test for NodeJS
Added Redux for communication between Settings component and other components.
Simplified/Generalized components with HOC

#### Who is reviewing it?
<!--(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)-->
@vinamartin @alchucam @mcalcote @haydenudelson
<!--(please choose AT LEAST one UI team reviewer)-->
@Bdthomson @adimka @samuelechu @djblue @gordocanchola

#### How should this be tested?
`yarn go` and `yarn server` in two separate terminals, OR
`make go`

#### Screenshots
<!--(if appropriate)-->
![ezgif com-video-to-gif (1)](https://user-images.githubusercontent.com/40583917/63979672-48511f00-ca6e-11e9-8db2-e997b71f0475.gif)
![ezgif com-video-to-gif (2)](https://user-images.githubusercontent.com/40583917/63979710-5d2db280-ca6e-11e9-96a0-fa91a2c2be33.gif)

#### Checklist:
- [x] Documentation Updated

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist.
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

vinamartin commented:
Remove the edit button from the media component as part of this change. (alternatively: hide it until you're in edit mode)

vinamartin commented on client/wallboards/I2OWallboard.js:
```diff
@@ -5,7 +5,12 @@ import Grid from "@material-ui/core/Grid";
+import { updateCurrentWallboard, leaveEdit } from "../actions";
 
+const buildStatusWidth = `100%`;
+// const buildStatusWidth = `calc(100% - ${CARD_SIDE_MARGINS}px)`;
```
✏️ Do you still need this?

vinamartin commented on server/__tests__/app.test.js:
```diff
@@ -0,0 +1,56 @@
+          "/theme?wallboard=TV&component=BANNER"
+        );
+        expect(response.body).toEqual(expect.not.objectContaining({}));
+        // expect(response.body).toEqual({data:"red"});
```
✏️ Still need this?

vinamartin commented on server/__tests__/app.test.js:
```diff
@@ -0,0 +1,56 @@
+
+describe("GET ", () => {
+  describe("/theme ", () => {
+    test("with empty query", async () => {
```
✏️ Can you make these slightly more descriptive ?

haydenudelson commented on client/components/Settings/componentHOC.js:
```diff
@@ -0,0 +1,402 @@
+    }
+
+    //second Dialog which appears once the user saved the input.
+    displaySaveResults() {
```
✏️ I think it would be better to have the feedback on the same dialog as the input.  It's a little annoying having to close out of two dialogs after you submit.  You could have a message appear above the save button.

alchucam replied:
now it close automatically with a timeout!

haydenudelson commented on client/components/Settings/componentHOC.js:
```diff
@@ -0,0 +1,402 @@
+            {" "}
+            {/* close the opened result popup */}
+            <CancelFabWrapper />
+            Cancel
```
✏️ "close"- cancel implies to me that clicking the button undoes the operation.

haydenudelson commented on client/components/Settings/componentHOC.js:
```diff
@@ -0,0 +1,402 @@
+    }
+
+    displayBanner() {
+      let editable = this.state.edit && this.props.currentWallboard === "HOME";
```
❓ Why not be able to edit the banner from any wallboard?

alchucam replied:
updated to be able to edit the banner from any wallboard!

haydenudelson commented on client/components/SlackCard.js:
```diff
@@ -291,16 +291,23 @@ class SlackCard extends React.Component {
-        : message.attachments[0].author_icon;
+    let author;
+    let avatar;
+    if (message != undefined) {
```
👍 Good change!

haydenudelson commented on client/reducers/currentWallboard.js:
```diff
@@ -0,0 +1,10 @@
+const currentWallboard = (state = "HOME", action) => {
+  switch (action.type) {
+    case "UP_WALL":
```
❓ UPDATE_WALLBOARD?

haydenudelson commented on client/reducers/currentComponents.js:
```diff
@@ -0,0 +1,12 @@
+const currentComponents = (state = [], action) => {
```
✏️ If we're not using ADD_COMP, DEL_COMP yet I don't think we should include it in the PR

haydenudelson commented on client/utils/Constants.js:
```diff
@@ -1,3 +1,8 @@
+import { createMuiTheme } from "@material-ui/core/styles";
+
+// banner
+export const BANNER_HEIGHT = 100;
```
✏️ If we adjust the banner height, we need to update the Connexta logo so its pseudo-centered again (same margin below/above lower-case letters).  It's probably best to make the logo a little smaller to make this work.

alchucam replied:
reverted back to old banner height

haydenudelson commented on client/utils/DefaultData.js:
```diff
@@ -0,0 +1,20 @@
+import { CX_DARK_BLUE } from "./Constants";
+import { jenkinsURLList, AFJenkinLink, AFURL, AFpipeline } from "./Link";
+const CHANNEL = process.env.SLACK_CHANNEL;
```
✏️ Why declare these consts up here if we're just gonna declare them once down below?

haydenudelson commented on server/app.js:
```diff
@@ -0,0 +1,250 @@
+app.use(express.static(mediaDir));
+const themeFileLocation = prod
+  ? "/eve/theme.json"
+  : path.join(process.cwd(), "eve/theme.json");
```
❓ maybe we should have all the !prod values use `path.join(process.cwd(), ...)` to be consistent

alchucam commented on client/components/BuildStatus.js:
```diff
@@ -88,28 +85,32 @@ class BuildStatus extends React.Component {
+        );
+      }
+      //fetch and update jenkins information for all team
+      await Promise.all(this.trashableRequestList)
```
I didn't change this part of the code. I think prettified made it that there was a change. You can check by hiding whatspaces changes. But to answer your question, you are getting a list of Jenkins URL to be used for fetching data, which contains weatherscore, etc.

alchucam commented:
New Update:
1. Edit button of Media, Event and Version components are now hidden while not in the edit mode
2. links of components are now disable while in edit mode
3. Update ReadMe to describe how to add new wallboard & new component

alchucam commented:
> I think you did a great job w this PR, Alex. It was a big undertaking, and you designed an elegant system to take care of the theming.
> 
> I had a couple general questions/notes:
> 
> * Should we make the media/event components editable as well.  When you click on them, you could be directed to the add/remove media and log in/select calendar dialogs.
> * We should disable links, buttons in the editing view.  For instance, when you click on the PR in the Github component to edit, it opens a new tab for the PR.  This isn't what we want.
> * It would be nice to have a dropdown for channel selection in Slack.  It's pretty hard to find a given channel's ID, which you need right now.
> * For Jenkins, should we just have the user inputted value not show up instead of displaying invalid? I can't think why you would want to see invalid displayed on the screen.  It might be nice to have the current URL/name values autofilled in those textfields as well.
> * Why can't we edit the I20 build status component?
> * Does it make sense to have the edit feature on the TV wallboard?  I think it would be best (maybe for another ticket) to have nothing interactive on the TV display, since that's a little visually confusing.
> * It would be cool if you could add/remove Jenkins builds instead of editing a given set of 5.
> 
> I don't think all these concerns should be addressed in this ticket, but I wanted to make note of them.

Thanks for the good feedback! I've addressed some of the feedback to this PR.
To answer some questions, 
1. I2O is currently disabled for edit as it is already a very specifically customized BuildStatus (also using TeamBuildStatus).
2. Hiding settings icons in TV wallboard is a good idea to answer your question though I believe it can be resolved in another ticket
3. Add/remove Jenkins builds are limited to 6 for TV buildStatus, 3 for GSR since we are currently setting the height and width of the components in the code; therefore, if we don't limit the number of URLs to be added,  the components will be broken with increased number which is more than the current width can handle.
4. Dropdown instead of typing is a great idea. It is possible to achieve as well by placing the all the extracted data into the dropdown selection; however, I believe it should be handled as an additional ticket since this PR is already quite huge.

haydenudelson commented:
✅ Hero'd!
